iocinf.cpp: Hostnames may be longer than 32 bytes

tboegi opened the pull request:
Found here at ESS:

(all in one line)
EPICS_PVA_ADDR_LIST=averylonghostname.mylabnetwork.technicalnetwork.example.com
EPICS_PVA_AUTO_ADDR_LIST=NO
pvget somePVnam

leads to something like this:
CA.Client.Exception...............................................
    Warning: "Empty PV search address list"
    Source File: ../udpiiu.cpp line 403
    Current Time: Thu Jun 09 2022 10:10:47.804161447

Solution:
Increase the buf size in addAddrToChannelAccessAddressList() from 32 to 256

anjohnson reviewed (approved):
I suspect the real length limit for a fully-qualified hostname is likely to be larger than 256, but I'm not suggesting we worry about that. This looks fine to me.

mdavidsaver commented:
If @anjohnson isn't worried about the increased stack usage, then I don't see a problem with this as a quick fix.  Although being c++, `std::vector` is available.

In the longer term, our general trend has been to redesign code to remove fixed size limits like this altogether.  eg. make a temporary copy in a mutable `std::vector<char>`, and then split with `epicsStrtok_r()`.

tboegi commented:
> I suspect the real length limit for a fully-qualified hostname is likely to be larger than 256, but I'm not suggesting we worry about that. This looks fine to me.
@anjohnson 
I think that we are fine with 256:
https://github.com/epics-base/epics-base/commit/dca5c9ed3c038e39c1e34e3dd14ca5633ac732d4

anjohnson commented:
@tboegi Thanks for the RFC pointer, it's nice to see that the length limit is actually well-defined and not huge.

tboegi committed:
iocinf.cpp: Hostnames may be longer than 32 bytes

Found here at ESS:

(all in one line)
EPICS_CA_ADDR_LIST=averylonghostname.mylabnetwork.technicalnetwork.example.com
EPICS_CA_AUTO_ADDR_LIST=NO
caget somePVnam

leads to something like this:
CA.Client.Exception...............................................
    Warning: "Empty PV search address list"
    Source File: ../udpiiu.cpp line 403
    Current Time: Thu Jun 09 2022 10:10:47.804161447

Problem desription:
addAddrToChannelAccessAddressList() will collect what ever we specify in
EPICS_CA_ADDR_LIST for channel access.
That function will add IP-addresses to the search list.
hostnames are possible, but are ignored if longer than 32 bytes,
because buf is too short.

If a hostname can be resolved into an IP, that is fine,
if not that is "fine as well" (better say: silently ignored):
If, and only if, EPICS_CA_AUTO_ADDR_LIST=NO is given then the one
and only too long hostname will be ignored and lead to an
"Empty PV search address list".

If EPICS_CA_AUTO_ADDR_LIST=YES (or nothing) is specified, the search list
is not empty, and EPICS will search all broadcast addresses.
This will eventually lead into a timout.

Solution:
Increase the buf size in addAddrToChannelAccessAddressList() from 32 to 256

The maximum length for a hostname is 255:
https://www.ietf.org/rfc/rfc1034.txt

If we add one byte for the string terminating '\0', we need 256 bytes.

And yes, this patch neglets the fact that a user can specify HOSTNAME:PORT,
which may need another 6 bytes. To put it the other way around:
In this case the hostname length is limited to 250 bytes,
which is still long enough in practice.

mdavidsaver commented:
Applied as a8e8d22c31aaa2bfff6511017e883b5e751a3e0a
